internal/caveats/debug.go

```diff
@@ -0,0 +1,149 @@
+	childExprStrings := make([]string, 0, len(sr.exprResultsForDebug))
+	combinedContext := map[string]any{}
+
+	for _, child := range sr.exprResultsForDebug {
+		if _, ok := child.(*caveats.CaveatResult); ok {
+			childExprString, contextMap, err := buildDebugInformationForConcrete(child.(*caveats.CaveatResult), resultsByParam)
+			if err != nil {
+				return "", nil, err
+			}
+
+			childExprStrings = append(childExprStrings, "("+childExprString+")")
+			maps.Copy(combinedContext, contextMap)
+			continue
+		}
+
+		childExprString, contextMap, err := buildDebugInformation(child.(syntheticResult), resultsByParam)
+		if err != nil {
+			return "", nil, err
+		}
+
+		childExprStrings = append(childExprStrings, "("+childExprString+")")
+		maps.Copy(combinedContext, contextMap)
+	}
+
+	var combinedExprString string
+	switch sr.op {
+	case corev1.CaveatOperation_AND:
+		combinedExprString = strings.Join(childExprStrings, " && ")
+
+	case corev1.CaveatOperation_OR:
+		combinedExprString = strings.Join(childExprStrings, " || ")
+
+	case corev1.CaveatOperation_NOT:
+		if len(childExprStrings) != 1 {
+			return "", nil, spiceerrors.MustBugf("NOT operator must have exactly one child")
+		}
+
+		combinedExprString = "!" + childExprStrings[0]
+
+	default:
+		return "", nil, fmt.Errorf("unknown operator: %v", sr.op)
+	}
+
+	return combinedExprString, combinedContext, nil
+}
+
+func buildDebugInformationForConcrete(cr *caveats.CaveatResult, resultsByParam *mapz.MultiMap[string, *caveats.CaveatResult]) (string, map[string]any, error) {
+	// For each paramter used in the context of the caveat, check if it is shared across multiple
+	// caveats. If so, rewrite the parameter to a unique name.
+	existingContextMap := cr.ContextValues()
+	contextMap := make(map[string]any, len(existingContextMap))
+
+	caveat := *cr.ParentCaveat()
+
+	for paramName, paramValue := range existingContextMap {
+		index := mapz.IndexOfValueInMultimap(resultsByParam, paramName, cr)
+		if resultsByParam.CountOf(paramName) > 1 {
+			newName := paramName + "__" + strconv.Itoa(index)
+			if resultsByParam.Has(newName) {
+				return "", nil, fmt.Errorf("failed to generate unique name for parameter: %s", newName)
+			}
+
+			rewritten, err := caveat.RewriteVariable(paramName, newName)
+			if err != nil {
+				return "", nil, err
+			}
+
+			caveat = rewritten
+			contextMap[newName] = paramValue
+			continue
+		}
+
+		contextMap[paramName] = paramValue
+	}
+
+	exprString, err := caveat.ExprString()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return exprString, contextMap, nil
+}
+
+func collectParameterUsage(sr syntheticResult, resultsByParam *mapz.MultiMap[string, *caveats.CaveatResult]) {
+	for _, exprResult := range sr.exprResultsForDebug {
+		if concrete, ok := exprResult.(*caveats.CaveatResult); ok {
+			for paramName := range concrete.ContextValues() {
+				resultsByParam.Add(paramName, concrete)
+			}
+		} else {
+			collectParameterUsage(exprResult.(syntheticResult), resultsByParam)
```

Comment:
Not a fan of unsafe type assertions, I'd rather return an error in this function, been when we know that _right now_ the error scenario is not possible. Code evolves over-time and makes this less obvious to catch.

Reply:
Changed